docs/src/main/asciidoc/security-openid-connect-client.adoc

```diff
@@ -0,0 +1,218 @@
+
+== Introduction
+
+`quarkus-oidc-client` extension provides a reactive `io.quarkus.oidc.client.OidcClient` which can be used to acquire and refresh tokens using Smallrye Mutiny `Uni` and `Vert.x WebClient`.
+
+`OidcClient` is initialized at the build time with the IDP token endpoint URL which can be auto-discovered or manually configured and uses this endpoint to acquire access tokens using `client_credentials` or `password` token grants and refresh the tokens using `refresh_token` grant.
+
+Here is how `OidcClient` can be configured to use the `client_credentials` grant:
+
+```
+quarkus.oidc.client.auth-server-url=${keycloak.url}/realms/quarkus2/
+quarkus.oidc.client.client-id=quarkus-app
+quarkus.oidc.client.credentials.secret=secret
+```
+
+Here is how `OidcClient` can be configured to use the `password` grant:
+
+```
+quarkus.oidc-client.auth-server-url=${keycloak.url}/realms/quarkus2/
+quarkus.oidc-client.client-id=quarkus-app
+quarkus.oidc-client.credentials.secret=secret
+quarkus.oidc-client.grant.type=user
+quarkus.oidc-client.grant-options.user.name=alice
+quarkus.oidc-client.grant-options.user.password=alice
+```
+
+Note the OAuth2 `password` grant is enabled with `quarkus.oidc-client.grant.type=user` to avoid duplicating `password` when setting a user password property required by the `password` grant.
+
+In both cases `OidcClient` will auto-discover the token endpoint URL and use it to acquire the tokens.
+
+== Use OidcClient directly
+
+One can use `OidcClient` directly as follows:
+
+```
+import javax.inject.PostConstruct;
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+
+import io.quarkus.oidc.client.OidcClient;
+import io.quarkus.oidc.client.GrantTokens;
+
+@Path("/service")
+public class OidcClientResource {
+
+    @Inject
+    OidcClient client;
+
+    volatile GrantTokens currentTokens;
+
+    @PostConstruct
+    public init() {
+        currentTokens = client.getTokens().await().indefinitely();
+    }
+
+    @GET
+    public String getResponse() {
+        
+        GrantTokens tokens = currentTokens;
+        if (tokens.isAccessTokenExpired) {
+            tokens = client.refreshTokens(tokens.getRefreshToken().await.indefinitely();
+            currentTokens = tokens;
+        } 
+        // use tokens.getAccessToken() to configure MP RestClient Authorization header/etc
+    }
+}
+```
+
+== Use OidcClient as Microprofile RestClient client filter
+
+`quarkus-oidc-rest-client` extension provides `io.quarkus.oidc.rest.client.OidcClientRequestFilter` JAX-RS ClientRequestFilter which uses `OidcClient` to acquire the access token, refresh it if needed, and set it as an HTTP `Authorization` `Bearer` scheme value.
+
+By default this filter will get `OidcClient` to acquire the first pair of access and refresh tokens at its initialization time. If the access tokens are short lived and refresh tokens are not available then the token acquisition should be delayed with `quarkus.oidc-client.early-tokens-acquisition=false`.
```

Comment:
I'm not sure it makes sense to set up the client with `user` because I'm not sure if that is a valid use case for chaining services. So, refresh tokens should only make sense if the application is a `web-app` that authenticates using code flow and needs to access downstream services.

For service-to-service communication it would makes more sense client credentials where refresh tokens are not supposed to be issued (in Keycloak we are changing this).

Btw, is it possible to use this filter using a bearer token sent to a `service` type application ? One thing nice here would be the possibility to automatically exchange tokens depending on the service the application is calling. So that another token is issued by the AS and meant to use in the target service.

Reply:
Hi @pedroigor yeah, I'm just trying to position this client as a generic oidc client, beyond chaining the services or a service to service, you know we have many places were we use RestAssured or Keycloak admin api to get a token with a password grant; and I'm also keeping an option open for supporting the custom grants.
Right now this filter is not linked to the incoming tokens, the case where the coming in bearer token is used in the next invocation is not covered - in this simplest form there is nothing we can suggest but using the context propagation where `JsonWebToken` is injected in the client filter where one would just add its value to the Authorization header - I think I may add this simple filter anyway - and we've had an exchange tokens issue closed recently till better times :-) - so after this PR, I'll reopen that issue and we can enhance it with the exchange protocol, that should be interesting

Reply:
Hey @pedroigor Re that new filter which would propagate the token, not sure I'll have time to put it into this PR, I'll see, but I'm thinking of naming it `TokenPropagationRequestFilter` as opposed to `OidcClientRequestFilter` which is shipped in this PR - the latter is not related in anyway to the endpoint authentication, while the former would pick up the injected token - in the 1st iteration it would just put it into the header, later - we will wire in the exchange protocol.
Does it sound good to you ?

Reply:
Sure